### PR TITLE
kube: proper/simplified syncing with client-go 1.27

### PR DESCRIPTION
**Pull request by howardjohn**
This extends https://github.com/istio/istio/pull/43870, with an upgrade to client-go 1.27 that introduces proper syncing. This was split out of that PR to avoid dependency update and more logic changes.

For the best details of what the "correct" way to sync is, see https://github.com/istio/istio/blob/4e05079e8f7ea1eb82df31ef99ddbfb0843dc213/pkg/kube/controllers/example_test.go#L88-L112.

Specifically, in 1.27 they allowed the ability to tell when a _event handler_ is synced, rather than the _informer store_. This is the critical piece we take advantage of.

This PR both fixes improper syncing in a bunch of controllers automatically (by the `kclient.Client` abstraction improving), and simplifies a few controllers that previously had correct (but complex) syncing.

Note this does upgrade to a pre-release of client-go. In T&R slack channel there was some consensus that this was worthwhile since the official release would be prior to 1.18.0

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
/test all

File: pilot/pkg/config/kube/crdclient/client.go [hunk 1]
```diff
@@ -239,34 +216,7 @@ func (cl *Client) informerSynced() bool {
...
-				if cl.objectInRevision(&currConfig) {
-					for _, f := range handlers {
-						f(config.Config{}, currConfig, model.EventAdd)
-					}
-				}
-			}
-		}()
-	}
-	wg.Wait()
+	return cl.queue.HasSynced()
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/client.go:219]: donot we need wait informer synced here?

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/client.go:219]: IC we wait informerSynced before run queue

**Review by hzxuzhonghu — APPROVED** [review]

**Review by richardwxn — APPROVED** [review]

**Pull request merged**